### ISSUE TITLE
More `Map` functions could use pointer equality to avoid allocations in the no-op case

### ISSUE THREAD

**Issue by sjakobi** [haskell/containers#836]
Currently the use of `ptrEq` for `Map` functions is limited to a select few functions defined in `Data.Map.Internal`. Notably it is not used at all in `Data.Map.Strict.Internal`.

In particular, I think that `adjust`, `update` and `alter` could use this optimization. But there are probably more candidates.

**Comment by treeowl** [comment]
Yes, the strict versions can do even more with pointer equality than the lazy ones.
